tasks/update_history.ps1

```diff
@@ -0,0 +1,81 @@
+Function Get-SafeString($value) {
+  if ($value -eq $null) {
+    Write-Output ''
+  } else {
+    Write-Output $value.ToString()
+  }
+}
+
+Function Get-SafeDateTime($value) {
+  if ($value -eq $null) {
+    Write-Output ''
+  } else {
+    Write-Output $value.ToString('u')
+  }
+}
+
+Function Convert-ToServerSelectionString($value) {
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa387280(v=vs.85).aspx
+  switch ($value) {
+    0 { Write-Output 'Default' }
+    1 { Write-Output 'ManagedServer' }
+    2 { Write-Output 'WindowsUpdate' }
+    3 { Write-Output 'Other' }
+    Default { Write-Output "Unknown ${value}"}
+  }
+}
+
+Function Convert-ToOperationResultCodeString($value) {
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa387095(v=vs.85).aspx
+  switch ($value) {
+    0 { Write-Output 'Not Started' }
+    1 { Write-Output 'In Progress' }
+    2 { Write-Output 'Succeeded' }
+    3 { Write-Output 'Succeeded With Errors' }
+    4 { Write-Output 'Failed' }
+    5 { Write-Output 'Aborted' }
+    Default { Write-Output "Unknown ${value}"}
+  }
+}
+
+Function Convert-ToUpdateOperationString($value) {
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa387282(v=vs.85).aspx
+  switch ($value) {
+    1 { Write-Output 'Installation' }
+    2 { Write-Output 'Uninstallation' }
+    Default { Write-Output "Unknown ${value}"}
+  }
+}
+
+$Session = New-Object -ComObject "Microsoft.Update.Session"
+$Searcher = $Session.CreateUpdateSearcher()
+# Returns IUpdateSearcher https://msdn.microsoft.com/en-us/library/windows/desktop/aa386515(v=vs.85).aspx
+
+$historyCount = $Searcher.GetTotalHistoryCount()
+$Searcher.QueryHistory(0, $historyCount) | ForEach-Object -Process {
+  # Returns IUpdateHistoryEntry https://msdn.microsoft.com/en-us/library/windows/desktop/aa386400(v=vs.85).aspx
+  $props = @{
+    'Operation' = Convert-ToUpdateOperationString $_.Operation
+    'ResultCode' = Convert-ToOperationResultCodeString $_.ResultCode
+    'HResult' = $_.HResult
+    'Date' = Get-SafeDateTime $_.Date
+    'UpdateIdentity' = @{}
+    'Title' = Get-SafeString $_.Title
+    'Description' = Get-SafeString $_.Description
+    'UnmappedResultCode' = $_.UnmappedResultCode
+    'ClientApplicationID' = Get-SafeString $_.ClientApplicationID
+    'ServerSelection' = Convert-ToServerSelectionString $_.ServerSelection
+    'ServiceID' = Get-SafeString  $_.ServiceID
+    'UninstallationSteps' = @()
+    'UninstallationNotes' = Get-SafeString $_.UninstallationNotes
+    'SupportUrl' = Get-SafeString $_.SupportUrl
+    'Categories' = @()
```

Comment:
Possibly better to set `Categories` equal to `New-Object System.Collections.ArrayList` and then on [line 75](https://github.com/puppetlabs/puppetlabs-wsus_client/pull/83/commits/4bcc461275a50e5bdb88e20fa8ca6ef2dbfd512f#diff-1b9bad9ab08b53a5e96ced4852470040R75) you could use `$props.Categories.Add($_.Name)`.

Reply:
The Categories is an enumerable COM object type.   By setting `Categories` to an array and then appending items to it, I can keep the type as `String[]`.  If I used a pipeline I will end up with `String` for single categories and `null` for empty categories.